plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -1574,14 +1571,11 @@ producer_plugin_impl::start_block_result producer_plugin_impl::start_block() {
                      } FC_LOG_AND_DROP();
                   }
```

Comment:
Unfortunately, this code will now drop unapplied transactions that are not expired and not persisted during speculative blocks. This only happened in the old code if there were no producers set and therefore no chance for these types of transactions to ever be applied (until their eventual expiration caused their removal).

I guess you could do a `_unapplied_transactions.add_subjective_failure( trx );` in an else branch to avoid dropping them. However, they would be put in the back of queue which is different behavior than what was happening in the old code (same goes for the `add_subjective_failure` call when the transaction is applied and fails with an exception that we consider subjective). Granted the old code ordered by transaction ID rather than using a FIFO, which is not great either, but I think we need to more carefully consider the implications of re-order the transactions each time we encounter them during a speculative block.

Honestly, I'm thinking a heavier implementation of `unapplied_transaction_queue` based on `multi_index` could be worth it in terms of the flexibility and hopefully developer simplicity it can buy us. It could also swallow up the currently separate `_pending_incoming_transactions` queue as well as hold a flag to record whether the transaction is persistent so that we wouldn't need `persisted_by_id` either. Then you could just have that queue (indexed with potentially multiple indices to provide the flexibility of choosing the next transaction that producer plugin needs to implement its specific scheduling policy) along with of course the deferred transaction queue (accompanied by `_blacklisted_transactions`) to implement the transaction scheduling policy in the producer plugin.